extractor/docker/docker.go

```diff
@@ -252,17 +234,70 @@ func (d DockerExtractor) Extract(ctx context.Context, imageName string, filename
 	if err != nil {
 		return nil, xerrors.Errorf("failed to decode config JSON: %w", err)
 	}
+	return config, nil
+}
 
-	// special file for command analyzer
-	fileMap["/config"] = config
+func (d DockerExtractor) extractLayerFiles(layerCh chan layer, errCh chan error, ctx context.Context, filenames []string, filesInLayers map[string]extractor.FileMap, opqInLayers map[string]extractor.OPQDirs) error {
+	var l layer
+	select {
+	case l = <-layerCh:
+	case err := <-errCh:
+		return err
+	case <-ctx.Done():
+		return xerrors.Errorf("timeout: %w", ctx.Err())
+	}
+	files, opqDirs, err := d.ExtractFiles(l.Content, filenames)
+	if err != nil {
+		return xerrors.Errorf("failed to extract files: %w", err)
+	}
+	layerID := string(l.ID)
+	filesInLayers[layerID] = files
+	opqInLayers[layerID] = opqDirs
 
-	return fileMap, nil
+	return nil
+}
+
+func (d DockerExtractor) extractLayerWorker(dig digest.Digest, r *registry.Registry, ctx context.Context, image registry.Image, errCh chan error, layerCh chan layer) {
+	var rc io.Reader
+	// Use cache
+	rc = d.Cache.Get(string(dig))
+	if rc == nil {
+		// Download the layer.
+		layerRC, err := r.DownloadLayer(ctx, image.Path, dig)
+		if err != nil {
+			errCh <- xerrors.Errorf("failed to download the layer(%s): %w", dig, err)
+			return
+		}
+
+		rc, err = d.Cache.Set(string(dig), layerRC)
+		if err != nil {
+			log.Print(err)
+		}
+	}
+	gzipReader, err := gzip.NewReader(rc)
+	if err != nil {
+		errCh <- xerrors.Errorf("invalid gzip: %w", err)
+		return
+	}
+	layerCh <- layer{ID: dig, Content: gzipReader}
+}
+
+func getValidManifest(err error, r *registry.Registry, ctx context.Context, image registry.Image) (*schema2.DeserializedManifest, error) {
```

Comment:
is `err error` necessary?

Reply:
nit: `ctx context.Context` is usually a first argument.

Reply:
Good catch. This is what happens when I rely on the IDE making the auto refactor for me 😅